File: neet_code_all_cpp/array_hash/first_missing_positive.cpp
#include <iostream>
#include <vector>

// 模範解答
class Solution {
    public:
        int firstMissingPositive(std::vector<int> nums) {
            for(int i = 0; i < nums.size(); ++i) {
                if(i + 1 == nums[i]) continue;
                int x = nums[i];
                while(x >= 1 && x <= nums.size() && x != nums[x - 1]) {
                    std::swap(x, nums[x - 1]);
                }
            }
            for(int i = 0; i < nums.size(); ++i) {
                if(i + 1 != nums[i]) {
                    return i + 1;
                }
            }

            return nums.size() + 1;
        }
};

int main(void) {
    std::vector<int> case_1 = {1, 2, 0};
    // => 3
    std::vector<int> case_2 = {3, 4, -1, 1};
    // => 2
    std::vector<int> case_3 = {7, 8, 9, 11, 12};
    // => 1

    Solution s_1;

    std::cout << s_1.firstMissingPositive(case_1) << std::endl;
    std::cout << s_1.firstMissingPositive(case_2) << std::endl;
    std::cout << s_1.firstMissingPositive(case_3) << std::endl;
}

File: neet_code_all/src/array_hash/first_missing_positive.rs
// 縛り: メモリ O(1), 時間 O(N)

// hit1: まずは追加のメモリのことを考えずに考えてみよう
// hit2: ダブった値や負の値は無視
// hit3: O(2n) = O(n)

use std::collections::HashMap;

// 解けなかった
struct Solution;
impl Solution {
    pub fn first_missing_positive(nums: Vec<i32>) -> i32 {
        let min_pos = nums.iter().fold(i32::MAX, |mut min, x| {
            if *x > 0 {
                min = min.min(*x);
            }
            min
        });
        let mut map: HashMap<i32, Vec<usize>> = HashMap::new();
        for (i, n) in nums.iter().enumerate() {
            if let Some(vec_n) = map.get_mut(&n) {
                vec_n.push(i);
            } else {
                map.insert(*n, vec![i]);
            }
        }
        println!("map: {:?}", map);

        1
    }
}

// C++の模範解答より
struct SolutionAnsCpp;
impl SolutionAnsCpp {
    // AC
    pub fn first_missing_positive(nums: Vec<i32>) -> i32 {
        let mut nums = nums;
        for i in 0..nums.len() {
            if i as i32 + 1 == nums[i] {
                continue;
            }
            let mut x = nums[i];
            while x >= 1 && x as usize <= nums.len() && x != nums[x as usize - 1] {
                let tmp = nums[x as usize - 1];
                nums[x as usize - 1] = x;
                x = tmp;
            }
        }

        for i in 0..nums.len() {
            if i as i32 + 1 != nums[i] {
                return i as i32 + 1;
            }
        }

        nums.len() as i32 + 1
    }
}

fn main() {
    let case_1 = vec![1, 2, 0];
    // => 3
    let case_2 = vec![3, 4, -1, 1];
    // => 2
    let case_3 = vec![7, 8, 9, 11, 12];
    // => 1

    /*
    println!(
        "case_1: {}",
        Solution::first_missing_positive(case_1.clone())
    );
    println!(
        "case_2: {}",
        Solution::first_missing_positive(case_2.clone())
    );
    */

    println!(
        "case_1: {}",
        SolutionAnsCpp::first_missing_positive(case_1.clone())
    );
    println!(
        "case_2: {}",
        SolutionAnsCpp::first_missing_positive(case_2.clone())
    );
    println!(
        "case_3: {}",
        SolutionAnsCpp::first_missing_positive(case_3.clone())
    );
}
